[REVIEW] Simplify tSNE perplexity search

zbjornson opened the pull request:
Sum(P) == n. See discussion in https://github.com/rapidsai/cuml/pull/1364#discussion_r453274117. This is also what CannyLab's implementation does.

While this removes code (including an atomic op), it somehow slightly increases runtime (~200 ms for 1M rows), but it's off the critical path and I think worth it for the 20 fewer lines of code and more accurate value. AFAICT it's due to the change to the `sum_disti_Pi` loop inside `sigmas_kernel`. (I'm inclined to guess it has to do with loop alignment, but nvcc doesn't provide a way to control that.)

@drobison00

GPUtester commented:
Can one of the admins verify this patch?

GPUtester commented:
Can one of the admins verify this patch?

zbjornson commented:
@cjnolet done!

zbjornson commented:
@cjnolet and/or @drobison00 could you review this when you get a chance please? I have some more tSNE fixes lined up that depend on this. I haven't made any changes to this PR since Jul, aside from rebasing it periodically.

cjnolet reviewed (approved):
@zbjornson, this looks great and I see a lot of potential to remove unnecessary memory reads/writes in these kernels by loading from distances once in each loop, performing the intermediate computations in registers and then making a single write to P at the very end. 

I'm approving this PR. Can you fix the conflicts?

zbjornson commented:
Thanks. Conflicts fixed.

zbjornson commented:
@cjnolet gentle nudge, this one's ready to merge I think!

JohnZed commented:
Just had a changelog conflict (fixed). I'll merge as soon as tests pass. Thanks!

zbjornson commented:
@JohnZed @cjnolet arghh, I just rebased again and the tests passed, but another changelog conflict arose while the tests were running. Just rebased again, but could this be merged without waiting for the tests?

JohnZed commented:
I'll merge today - sorry about the delay, @zbjornson
